File: COBRA_COMMANDER/comet.py
"""Abstraction layer on top of the DMX interface to a Coemar Comet."""
from collections import deque
from utils import unit_float_to_range, ignore_all_but_1, quadratic_fader
import logging as log

class Comet(object):
    """Better control abstraction for the Coemar Comet.

    Shutter control precedence (highest to lowest):
        shutter_open?
        shutter_sound_active?
        strobing? (with rate)

    Trigger control precendence:
        music_trigger?
        auto_step?
        manual step trigger action?
    """

    _game_dmx_vals = [12, 35, 65, 85, 112, 140, 165, 190, 212, 240]

    def __init__(self, dmx_addr):
        """Create a new wrapper for a comet."""
        self.dmx_addr = dmx_addr - 1

        self.shutter_open = False
        self.strobing = False
        self.strobe_rate = 0.0
        self.shutter_sound_active = False

        self.macro_pattern = 0
        self.mirror_speed = 0.0

        self.trigger_state = TriggerState()

        self.reset = False

    def update(self, timestep):
        self.trigger_state.update(timestep)

    def render(self, dmx_univ):
        """Render this Comet into a DMX universe."""

        # render the shutter value
        dmx_univ[self.dmx_addr] = self._render_shutter()
        dmx_univ[self.dmx_addr + 1] = self._game_dmx_vals[self.macro_pattern]
        dmx_univ[self.dmx_addr + 2] = self._render_mspeed()
        dmx_univ[self.dmx_addr + 3] = self.trigger_state.render()
        # reset
        dmx_univ[self.dmx_addr + 4] = 255 if self.reset else 0

        log.debug(dmx_univ[self.dmx_addr:self.dmx_addr+5])

    def _render_shutter(self):
        """Render the shutter state into DMX."""
        if not self.shutter_open:
            return 0
        elif self.shutter_sound_active:
            return 125
        elif self.strobing:
            return unit_float_to_range(151, 255, self.strobe_rate)
        else:
            return 75

    def _render_mspeed(self):
        return unit_float_to_range(0, 255, self.mirror_speed)

# Idle isn't a great name, as the comet could be in music trig or auto trig
# the point here is that the Idle state is any state that implies a DMX
# value outside the take a step ranges
Idle, SteppingForwards, SteppingBackwards = "I", "SF", "SB"
Forward, Backward = "F", "B"

class TriggerState(object):
    """Helper object to contain Comet trigger state."""
    # rendering to the enttec is asynchronous and frame tearing is a problem
    # how many updates should we hold the current output before processing another?
    _updates_to_hold = 3

    _step_forward_dmx_val = 108
    _step_backward_dmx_val = 142
    _stop_dmx_val = 124
    _music_dmx_val = 50

    def __init__(self):

        # stateless quantities
        self.music_trigger = False
        self.auto_step_rate = 0.0
        self.auto_step = False

        # queue of step actions to process
        self.steps_to_take = deque()

        # what state was this machine in on the last update?
        self.prior_state = Idle

        # how many frames should we hold current output before re-updating?
        self.updates_to_hold = 0

        self.current_output_value = self._stop_dmx_val

    def step_forwards(self):
        self.steps_to_take.appendleft(Forward)

    def step_backwards(self):
        self.steps_to_take.appendleft(Backward)

    def render(self):
        return self.current_output_value

    def update(self, _):
        if self.updates_to_hold > 0:
            self.updates_to_hold -= 1
        else:
            self.current_output_value = self._update()

    def _update(self):
        """Update the DMX trigger state value.

        This is a fairly complex action, as the step interface at the DMX level
        is kinda hokey.

        I give top priority to the mechanism by which we achieve manual stepping.
        After that, music responsive mode.  After that, automatic stepping.
        The trigger UI should make it clear that music trigger and auto trigger
        are mutually exclusive options.
        """
        # what needs to happen to take a step:
        # the dmx value needs to go from its current state to the step value
        # if the current value is the step value, we need to leave and come back again
        if self.steps_to_take:

            next_step = self.steps_to_take[-1]
            # hold this output for a minimum frame count before processing next
            self.updates_to_hold += self._updates_to_hold - 1

            if next_step == Forward and self.prior_state != SteppingForwards:
                # can take this step, transition to forward
                self.prior_state = SteppingForwards
                self.steps_to_take.pop()
                return self._step_forward_dmx_val
            elif next_step == Backward and self.prior_state != SteppingBackwards:
                # can take this step, transition to backward
                self.prior_state = SteppingBackwards
                self.steps_to_take.pop()
                return self._step_backward_dmx_val

            else:
                # we're in the same state as the step we need to take
                # transition to Idle, then take the step on the next update
                self.prior_state = Idle
                return self._stop_dmx_val

        # nothing in the step queue so the state machine is idle
        self.prior_state = Idle

        # if we're not taking a step, easy sauce
        if self.music_trigger:
            return self._music_dmx_val
        elif self.auto_step:
            return unit_float_to_range(151, 255, self.auto_step_rate)
        else:
            return self._stop_dmx_val

# controls and control actions

(Shutter,
 Strobe,
 StrobeRate,
 ShutterSoundActive,
 SelectMacro,
 Mspeed,
 Reset,
 StepForwards,
 StepBackwards,
 TrigSoundActive,
 AutoStep,
 AutoStepRate) = range(12)

# control actions
def shutter_state(comet, state):
    """bool"""
    comet.shutter_open = state

def strobe_state(comet, state):
    """bool"""
    comet.strobing = state

def strobe_rate(comet, rate):
    """float in [0,1]"""
    comet.strobe_rate = rate

def shutter_sound_active(comet, state):
    """bool"""
    comet.shutter_sound_active = state

def select_macro_pattern(comet, pattern):
    """int in [0,9]"""
    comet.macro_pattern = pattern

def mspeed(comet, mspeed):
    """float in [0,1]"""
    comet.mirror_speed = mspeed

def reset(comet, reset):
    """bool"""
    comet.reset = reset

def step_forwards(comet, _):
    comet.trigger_state.step_forwards()

def step_backwards(comet, _):
    comet.trigger_state.step_backwards()

def trigger_sound_active(comet, state):
    """bool"""
    comet.trigger_state.music_trigger = state

def auto_step(comet, state):
    """bool"""
    comet.trigger_state.auto_step = state

def auto_step_rate(comet, rate):
    """float in [0,1]"""
    comet.trigger_state.auto_step_rate = rate


# control mapping
control_map = {
    Shutter: shutter_state,
    Strobe: strobe_state,
    StrobeRate: strobe_rate,
    ShutterSoundActive: shutter_sound_active,
    SelectMacro: select_macro_pattern,
    Mspeed: mspeed,
    Reset: reset,
    StepForwards: step_forwards,
    StepBackwards: step_backwards,
    TrigSoundActive: trigger_sound_active,
    AutoStep: auto_step,
    AutoStepRate: auto_step_rate,}


def setup_controls(cont):

    # make groups
    cont.create_control_group('Controls')
    cont.create_control_group('Music')
    cont.create_control_group('Debug')

    # add controls
    cont.create_simple_control('Controls', 'Shutter', Shutter)
    cont.create_simple_control('Controls', 'Strobe', Strobe)
    cont.create_simple_control('Controls', 'StrobeRate', StrobeRate, quadratic_fader)
    cont.create_simple_control('Controls', 'Mspeed', Mspeed)
    cont.create_simple_control('Controls', 'AutoStep', AutoStep)
    cont.create_simple_control('Controls', 'AutoStepRate', AutoStepRate)

    cont.create_simple_control('Controls', 'StepBackwards', StepBackwards, ignore_all_but_1)
    cont.create_simple_control('Controls', 'StepForwards', StepForwards, ignore_all_but_1)

    cont.create_radio_button_control('Controls', 'SelectMacro', (10,1), SelectMacro)

    cont.create_simple_control('Music', 'ShutterSoundActive', ShutterSoundActive)
    cont.create_simple_control('Music', 'TrigSoundActive', TrigSoundActive)

    cont.create_simple_control('Debug', 'Reset', Reset)

File: rust/src/comet.rs
use log::{debug, error};
use number::UnipolarFloat;
use std::{collections::VecDeque, time::Duration};

use crate::dmx::DmxAddr;

pub struct Comet {
    dmx_addr: DmxAddr,
    shutter_open: bool,
    strobing: bool,
    strobe_rate: UnipolarFloat,
    shutter_sound_active: bool,
    macro_pattern: usize,
    mirror_speed: UnipolarFloat,
    trigger_state: TriggerState,
    reset: bool,
}

impl Comet {
    const GAME_DMX_VALS: [u8; 10] = [12, 35, 65, 85, 112, 140, 165, 190, 212, 240];

    pub fn new(dmx_addr: DmxAddr) -> Self {
        Self {
            dmx_addr,
            shutter_open: false,
            strobing: false,
            strobe_rate: UnipolarFloat::ZERO,
            shutter_sound_active: false,
            macro_pattern: 0,
            mirror_speed: UnipolarFloat::ZERO,
            trigger_state: TriggerState::new(),
            reset: false,
        }
    }

    pub fn update(&mut self, delta_t: Duration) {
        self.trigger_state.update(delta_t);
    }

    /// Render into the provided DMX universe.
    pub fn render(&self, dmx_univ: &mut [u8]) {
        dmx_univ[self.dmx_addr] = self.render_shutter();
        dmx_univ[self.dmx_addr + 1] = Self::GAME_DMX_VALS[self.macro_pattern];
        dmx_univ[self.dmx_addr + 2] = self.render_mspeed();
        dmx_univ[self.dmx_addr + 3] = self.trigger_state.render();
        dmx_univ[self.dmx_addr + 4] = if self.reset { 255 } else { 0 };
        debug!("{:?}", &dmx_univ[self.dmx_addr..self.dmx_addr + 5]);
    }

    fn render_shutter(&self) -> u8 {
        if !self.shutter_open {
            0
        } else if self.shutter_sound_active {
            125
        } else if self.strobing {
            unit_float_to_range(151, 255, self.strobe_rate)
        } else {
            75
        }
    }

    fn render_mspeed(&self) -> u8 {
        unit_float_to_range(0, 255, self.mirror_speed)
    }

    pub fn control(&mut self, msg: ControlMessage) {
        use ControlMessage::*;
        match msg {
            Set(sc) => self.handle_state_change(sc),
            Step(direction) => self.trigger_state.enqueue_step(direction),
        }
    }

    fn handle_state_change(&mut self, sc: StateChange) {
        use StateChange::*;
        match sc {
            Shutter(v) => self.shutter_open = v,
            Strobe(v) => self.strobing = v,
            StrobeRate(v) => self.strobe_rate = v,
            ShutterSoundActive(v) => self.shutter_sound_active = v,
            SelectMacro(v) => {
                if v >= Self::GAME_DMX_VALS.len() {
                    error!("Macro index {} out of range.", v);
                    return;
                }
                self.macro_pattern = v;
            }
            MirrorSpeed(v) => self.mirror_speed = v,
            TrigSoundActive(v) => self.trigger_state.music_trigger = v,
            AutoStep(v) => self.trigger_state.auto_step = v,
            AutoStepRate(v) => self.trigger_state.auto_step_rate = v,
            Reset(v) => self.reset = v,
        };
    }
}

/// Manage Comet trigger state.
struct TriggerState {
    music_trigger: bool,
    auto_step_rate: UnipolarFloat,
    auto_step: bool,
    /// queue of step actions to process
    steps_to_take: std::collections::VecDeque<Step>,
    /// what state was this machine in on the last update?
    prior_state: Stepping,
    /// how many frames should we hold current output before re-updating?
    updates_to_hold: usize,
    current_output_value: u8,
}

impl TriggerState {
    const DMX_VAL_STEP_FORWARD: u8 = 108;
    const DMX_VAL_STEP_BACKWARD: u8 = 142;
    const DMX_VAL_STOP: u8 = 124;
    const DMX_VAL_MUSIC_TRIG: u8 = 50;

    /// rendering to the enttec is asynchronous and frame tearing is a problem
    /// how many updates should we hold the current output before processing another?
    const UPDATES_TO_HOLD: usize = 3;

    fn new() -> Self {
        Self {
            music_trigger: false,
            auto_step_rate: UnipolarFloat::ZERO,
            auto_step: false,
            steps_to_take: VecDeque::new(),
            prior_state: Stepping::Idle,
            updates_to_hold: 0,
            current_output_value: Self::DMX_VAL_STOP,
        }
    }

    fn enqueue_step(&mut self, direction: Step) {
        self.steps_to_take.push_back(direction);
    }

    fn render(&self) -> u8 {
        self.current_output_value
    }

    pub fn update(&mut self, _: Duration) {
        if self.updates_to_hold > 0 {
            self.updates_to_hold -= 1;
        } else {
            self.current_output_value = self._update();
        }
    }

    /// Update the DMX trigger state value.
    ///
    /// This is a fairly complex action, as the step interface at the DMX level
    /// is kinda hokey.
    ///
    /// I give top priority to the mechanism by which we achieve manual stepping.
    /// After that, music responsive mode.  After that, automatic stepping.
    /// The trigger UI should make it clear that music trigger and auto trigger
    /// are mutually exclusive options.
    fn _update(&mut self) -> u8 {
        // what needs to happen to take a step:
        // the dmx value needs to go from its current state to the step value
        // if the current value is the step value, we need to leave and come back again
        if let Some(next_step) = self.steps_to_take.pop_front() {
            // hold this output for a minimum frame count before processing next
            self.updates_to_hold += Self::UPDATES_TO_HOLD - 1;

            if next_step == Step::Forward && self.prior_state != Stepping::Forwards {
                // can take this step, transition to forward
                self.prior_state = Stepping::Forwards;
                return Self::DMX_VAL_STEP_FORWARD;
            } else if next_step == Step::Backward && self.prior_state != Stepping::Backwards {
                // can take this step, transition to backward
                self.prior_state = Stepping::Backwards;
                return Self::DMX_VAL_STEP_BACKWARD;
            } else {
                // we're in the same state as the step we need to take
                // transition to Idle, then take the step on the next update
                self.steps_to_take.push_front(next_step);
                self.prior_state = Stepping::Idle;
                return Self::DMX_VAL_STOP;
            }
        }

        // nothing in the step queue so the state machine is idle
        self.prior_state = Stepping::Idle;

        // if we're not taking a step, easy sauce
        if self.music_trigger {
            return Self::DMX_VAL_MUSIC_TRIG;
        } else if self.auto_step {
            return unit_float_to_range(151, 255, self.auto_step_rate);
        } else {
            return Self::DMX_VAL_STOP;
        }
    }
}

#[derive(PartialEq)]
enum Stepping {
    Idle,
    Forwards,
    Backwards,
}

#[derive(PartialEq)]
pub enum Step {
    Forward,
    Backward,
}

pub enum ControlMessage {
    Set(StateChange),
    Step(Step),
}

pub enum StateChange {
    Shutter(bool),
    Strobe(bool),
    StrobeRate(UnipolarFloat),
    ShutterSoundActive(bool),
    SelectMacro(usize),
    MirrorSpeed(UnipolarFloat),
    TrigSoundActive(bool),
    AutoStep(bool),
    AutoStepRate(UnipolarFloat),
    Reset(bool),
}
